### ISSUE TITLE
[Proposal] Experimental mutable Failable

### ISSUE THREAD

**Issue by kourge** [UrbanDoor/failable#14]
We've been experimenting with using MobX and it definitely shows that Failable has more room for a MobX-friendly API. The proposal is thus: make Failable more object-oriented and `@observable` friendly.

```ts
declare class Failable<T> {
  readonly isSuccess: boolean;
  readonly isFailure: boolean;
  readonly isPending: boolean;

  success(data: T): this;
  failure(error: Error): this;
  pending(): this;

  match<A, B, C>(options: Failable.MatchOptions<T, A, B, C>): A | B | C;

  map<U>(f: (in: T) => U): Failable<U>;
  rescue(f: (error: Error) => T): this;
}

declare namespace Failable {
  export interface MatchOptions<T, A, B, C> {
    success: (data: T) => A;
    failure: (error: Error) => B;
    pending: () => C;
  }
}
```

An example usage would be:
```ts
const magicNumber = new Failable<number>();
// magicNumber.isPending === true;

// declare function fetchMagic(): Promise<number>;
fetchMagic().then(
  magicNumber.success,
  magicNumber.failure
);

// some time later:
const normalizedMagicPlusOne = magicNumber.match({
  pending: () => -1,
  failure: (_error) => 0,
  success: i => i + 1
});
```

Another advantage this gives us is subclassability. For example, if we wanted to log every error that's funneled through a Failable, simply subclass Failable and use that instead:

```ts
export class LoggedFailable<T> extends Failable<T> {
  failure(error: Error): this {
    super.failure(error);
    Raven.logException(error);
  }
}
```

**Comment by squirly** [comment]
I think this a great idea!

**Comment by kourge** [comment]
I'm soliciting more feedback on two other possible abstractions. Take a look at [`fromPromise` in `mobx-utils`](https://github.com/mobxjs/mobx-utils/blob/master/README.md#frompromise). We can more or less replicate that functionality with the following code:

```ts
function fromPromise<T>(p: PromiseLike<T>): Failable<T> {
  const f = new Failable<T>();
  p.then(f.success, f.failure);
  return f;
}
```

This means in comparison to `fromPromise`, `Failable<T>` sits slightly lower in terms of abstraction level. From a composability standpoint, however, `fromPromise` is less useful. Suppose we have a function that already fetches a user and gives back a promise:

```ts
declare function getUser(id: string): Promise<User>;
```

Let's start with immutable `Failable`, and how you'd manage fetches with it in a store:

```ts
class UserStore {
  @observable user: Failable<User> = pending;

  @action update = (id: string) => {
    getUser(id).then(user => {
      this.user = success(user);
    }, error => {
      this.user = failure(error);
    });
  };
}
```

With a mutable `Failable`, assuming the API suggestion in the top post, this becomes much simpler:

```ts
class UserStore {
  @observable user = new Failable<User>();

  @action update = (id: string) => {
    getUser(id).then(this.user.success, this.user.failure);
  };
}
```

If we were to use `fromPromise`, it would in turn look like this:

```ts
class UserStore {
  @observable user = new Failable<User>();

  @action update = (id: string) => {
    this.user = fromPromise(getUser(id));
  };
}
```

This continues to work, but only if someone is observing a `UserStore` instance as a whole. Let `store: UserStore`. In MobX parlance, for the updates to be propagated, you need to `observe(store)` or `observe(store, 'user')`. Simply doing `observe(store.user)` will not work. (See the docs on [`observe`](https://mobx.js.org/refguide/observe.html#observe) for a more in-depth explanation as to how these things are different.)

My question is: is this a big enough problem that we want to tackle it? I imagine a solution to this looks like:

```ts
class SourcedFailable<T> extends Failable<T> {
  absorb(promise: PromiseLike<T>): this {
    promise.then(this.success, this.failure);
    return this;
  }
}

class UserStore {
  @observable user = new SourcedFailable<User>();

  @action update = (id: string) => {
    this.user.absorb(getUser(id));
  };
}
```

**Comment by adamduffy** [comment]
what is the usage in the ui like?  are we still doing 

```
Failable.when(user, {
  pending: () => ...,
  failure: error => ...,
  success: user => ...
}```

maybe i missed this part somewhere?

**Comment by kourge** [comment]
The usage would look very similar:

```ts
store.user.match({
  pending: () => ...,
  failure: error => ...,
  success: user => ...
})
```

**Comment by kourge** [comment]
Closed via #17.
